### PR TITLE
Log MemoryUsed by transactions. 

### PR DESCRIPTION
**Pull request by dsainati1**
Add `MemoryUsed` to the transaction process, and then log it afterwards so we can pick it up in Grafana.

### CODE REVIEW

File: fvm/transaction.go [hunk 1]
```diff
@@ -77,6 +78,8 @@ func (proc *TransactionProcedure) Run(vm *VirtualMachine, ctx Context, st *state
 		}
 	}
 
+	proc.MemoryUsed = uint64(st.State().TotalMemoryUsed())
```
Comment (janezpodhostnik) [inline fvm/transaction.go:81]: can you move this to `transactionInvoker` and duplicate the logic for computation used:
- https://github.com/onflow/flow-go/blob/8f7942581efe7d86f7f7121a7cb6c1daf9fd3499/fvm/transactionInvoker.go#L248
- https://github.com/onflow/flow-go/blob/8f7942581efe7d86f7f7121a7cb6c1daf9fd3499/fvm/transactionInvoker.go#L167

That way we have that logic together.

**Review by janezpodhostnik — CHANGES_REQUESTED** [review]

**Review by janezpodhostnik — APPROVED** [review]
LGTM

File: engine/execution/computation/computer/computer.go [hunk 2]
```diff
@@ -407,19 +405,18 @@ func (e *blockComputer) executeTransaction(
...
+	runtime.ReadMemStats(&m)
+	memAllocAfter := m.TotalAlloc
+
 	evt := e.log.With().
 		Hex("tx_id", txResult.TransactionID[:]).
 		Str("block_id", res.ExecutableBlock.ID().String()).
 		Str("traceID", traceID).
 		Uint64("computation_used", txResult.ComputationUsed).
+		Uint64("memory_used", tx.MemoryUsed).
+		Uint64("memAlloc", memAllocAfter-memAllocBefore).
```
Comment (SupunS) [inline engine/execution/computation/computer/computer.go:417]: not something introduced in this PR, and probably not a big deal: wouldn't it be nice to use snake-case for all names for consistency? e.g: `memory_alloc`

**Review by SupunS — APPROVED** [review]
Is it possible to add a test-case?

File: fvm/fvm_test.go [hunk 3]
```diff
@@ -3539,6 +3539,7 @@ func TestSettingExecutionWeights(t *testing.T) {
 			tx := fvm.Transaction(txBody, 0)
 			err = vm.Run(ctx, tx, view, programs)
 			require.NoError(t, err)
+			require.Greater(t, tx.MemoryUsed, uint64(20_000_000_000))
```
Comment (dsainati1) [inline fvm/fvm_test.go:3542]: @SupunS should check that the `MemoryUsed` field is properly set/updated.

File: fvm/transactionInvoker.go [hunk 4]
```diff
@@ -246,6 +247,7 @@ func (i *TransactionInvoker) Process(
 	// if tx failed this will only contain fee deduction logs
 	proc.Logs = append(proc.Logs, env.Logs()...)
 	proc.ComputationUsed = proc.ComputationUsed + computationUsed
+	proc.MemoryUsed = proc.MemoryUsed + memoryUsed
```
[the remark this answers is not in the document]
Comment (dsainati1) [reply fvm/transactionInvoker.go:250]: If our memory limit is lower than the max value for an integer, we should never have to worry about this overflowing right? The transaction should abort long before we near the point of overflow.

**Comment by SaveTheRbtz** [comment]
Please note that `runtime.ReadMemStats` is a very expensive function to call since it calls `stopTheWorld` and waits for all goroutines to park:
```
// stopTheWorld stops all P's from executing goroutines, interrupting
// all goroutines at GC safe points and records reason as the reason
// for the stop. On return, only the current goroutine's P is running.
...
```

**Pull request merged**